### ISSUE TITLE
"TypeError: this._moduleRef.injector.get(...).runInitializers is not a function" when using BrowserModule.withServerTransition

### ISSUE THREAD

**Issue by mihalcan** [getsaf/shallow-render#83]
Test run fails with the following error when using Angular Universal with an import `BrowserModule.withServerTransition({ appId: 'serverApp' })` :
```
TypeError: this._moduleRef.injector.get(...).runInitializers is not a function
            at TestBedViewEngine.push.../../node_modules/@angular/core/fesm5/testing.js.TestBedViewEngine._initIfNeeded (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/@angular/core/fesm5/testing.js?:1702:59)
            at TestBedViewEngine.push.../../node_modules/@angular/core/fesm5/testing.js.TestBedViewEngine.createComponent (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/@angular/core/fesm5/testing.js?:1861:1)
            at Function.push.../../node_modules/@angular/core/fesm5/testing.js.TestBedViewEngine.createComponent (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/@angular/core/fesm5/testing.js?:1554:1)
            at Renderer.<anonymous> (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/shallow-render/dist/lib/models/renderer.js?:149:1)
            at step (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/shallow-render/dist/lib/models/renderer.js?:50:1)
            at Object.next (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/shallow-render/dist/lib/models/renderer.js?:31:45)
            at fulfilled (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/shallow-render/dist/lib/models/renderer.js?:22:42)
            at ZoneDelegate.../../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/zone.js/dist/zone.js?:388:1)
            at ProxyZoneSpec.push.../../node_modules/zone.js/dist/zone-testing.js.ProxyZoneSpec.onInvoke (http://localhost:9876/C:/GIT/game-monorepo/game/node_modules/zone.js/dist/zone-testing.js?:288:1)
```

Stackblitz example https://stackblitz.com/edit/github-ccz1rn?file=examples%2Fmodule-with-server-transition.spec.ts

Please do let me know if you need any help with the investigation.

**Comment by mihalcan** [comment]
I created the following `ExtendedShallow` class which just replaces the declaration of BrowserModule:
```Typescript
import { NgModule, Type } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { Shallow } from 'shallow-render';
import { getNgModuleAnnotations } from 'shallow-render/dist/lib/tools/get-ng-module-annotations';

function replaceBrowserModuleDeclaration(imports: any[]) {
  const idx = imports.findIndex(
    (imp: any) =>
      imp.ngModule && imp.ngModule.toString().indexOf('BrowserModule') >= 0
  );
  return idx >= 0
    ? Object.assign([], imports, { [idx]: BrowserModule })
    : imports;
}

function addFeatures(originalModule: Type<any>): Type<any> {
  const moduleAnnotations = getNgModuleAnnotations(originalModule);
  const imports = replaceBrowserModuleDeclaration(moduleAnnotations.imports);

  @NgModule({
    ...moduleAnnotations,
    imports: [...imports]
  })
  class ExtendedTestModule {}

  return ExtendedTestModule;
}

export class ExtendedShallow<TTestComponent> extends Shallow<TTestComponent> {
  constructor(testComponent: Type<TTestComponent>, testModule: Type<any>) {
    super(testComponent, addFeatures(testModule));
  }
}
```

It seems that just replacing BrowserModule declaration fixes the problem. Hope it helps.

**Comment by getsaf** [comment]
Man, this was tough. Shallow has this bit where we tell it to `neverMock` the `BrowserModule` [here](https://github.com/getsaf/shallow-render/blob/master/lib/shallow.ts#L14).

Looks like that doesn't pick up on the `BrowserModule.withServerTansition`.

My quick fix was to use either:

```typescript
new shallow(FooComponent, FooModule)
  .replaceModule(BrowserModule, BrowserModule);
```
--OR--
```typescript
Shallow.alwaysReplaceModule(BrowserModule, BrowserModule);
```

Your code snippet helped out here. I think I have a bug in module matching when it comes to `NgModuleWithProviders`. I'll see if I can get some time to look into it unless you're up to it.

Probably somewhere in [here](https://github.com/getsaf/shallow-render/blob/master/lib/tools/ng-mock.ts#L31) we need to check to see if it's an `NgModuleWithProviders`.

**Comment by getsaf** [comment]
This should be fixed now (see: #84). Give [v7.1.2](https://github.com/getsaf/shallow-render/releases/tag/v7.1.2) a try.
